File: packages/contracts/src/machine-plugin.ts
import * as z from "zod";

import { MachineDefinitionSchema } from "./domain";
import {
  DirectionUnitSchema,
  FiniteNumberSchema,
  TransformSchema,
  UuidSchema,
  Vec3MmSchema,
} from "./primitives";

export const FiveAxisModeSchema = z.enum(["3plus2", "simultaneous-5axis"]);

/** Data-only description; not executable code or proof of solver support. */
export const MachinePluginSchema = z.strictObject({
  contractVersion: z.literal(1),
  pluginId: z.string().min(1).max(128).regex(/^[a-z0-9]+(?:[.-][a-z0-9]+)*(?![\s\S])/),
  pluginVersion: z.string().regex(/^\d+\.\d+\.\d+(?![\s\S])/).max(32),
  architecture: z.enum(["table-table", "head-table", "head-head"]),
  machine: MachineDefinitionSchema,
  capabilities: z.strictObject({
    modes: z.array(FiveAxisModeSchema).min(1).max(2),
    tcp: z.enum(["unsupported", "supported"]),
  }),
  // Chains run from the machine base to the attachment. Empty means fixed base.
  toolChainAxisIds: z.array(UuidSchema).max(5),
  workpieceChainAxisIds: z.array(UuidSchema).max(5),
  toolMount: z.strictObject({
    positionMm: Vec3MmSchema,
    toolAxisUnit: DirectionUnitSchema,
  }),
  workpieceMount: TransformSchema,
}).superRefine((plugin, context) => {
  const issue = (path: (string | number)[], message: string) =>
    context.addIssue({ code: "custom", path, message });
  const { machine } = plugin;
  if (!["vertical-machining-center", "horizontal-machining-center"].includes(machine.machineType)) {
    issue(["machine", "machineType"], "five-axis milling requires a machining center");
  }
  if (machine.axes.length !== 5 || machine.axes.filter((axis) => axis.kind === "rotary").length !== 2) {
    issue(["machine", "axes"], "plugin requires exactly three linear and two rotary axes");
  }
  if (new Set(plugin.capabilities.modes).size !== plugin.capabilities.modes.length) {
    issue(["capabilities", "modes"], "modes must be unique");
  }
  const axes = new Map(machine.axes.map((axis) => [axis.id, axis]));
  const used = new Set<string>();
  for (const field of ["toolChainAxisIds", "workpieceChainAxisIds"] as const) {
    let parentId: string | null = null;
    plugin[field].forEach((id, index) => {
      const axis = axes.get(id);
      if (!axis || axis.parentId !== parentId || used.has(id)) {
        issue([field, index], "chains must be disjoint, known, root-to-leaf parent sequences");
      }
      used.add(id);
      parentId = id;
    });
  }
  if (machine.axes.some((axis) => !used.has(axis.id))) {
    issue(["machine", "axes"], "every machine axis must belong to exactly one attachment chain");
  }
  const headRotaries = plugin.toolChainAxisIds.filter((id) => axes.get(id)?.kind === "rotary").length;
  const expectedHeadRotaries = { "table-table": 0, "head-table": 1, "head-head": 2 }[plugin.architecture];
  if (headRotaries !== expectedHeadRotaries) {
    issue(["architecture"], "architecture must match rotary ownership of tool and workpiece chains");
  }
});

export type MachinePlugin = z.infer<typeof MachinePluginSchema>;
export type FiveAxisMode = z.infer<typeof FiveAxisModeSchema>;

/** Workpiece coordinates; tool-axis roll is intentionally not part of the task. */
export const FiveAxisTargetSchema = z.strictObject({
  tcpPositionMm: Vec3MmSchema,
  toolAxisUnit: DirectionUnitSchema,
});
export type FiveAxisTarget = z.infer<typeof FiveAxisTargetSchema>;
export type MachinePluginState = z.infer<ReturnType<typeof createMachinePluginStateSchema>>;

export const SolutionWeightsSchema = z.strictObject({
  axisTravel: z.number().int().min(0).max(1000),
  limitPenalty: z.number().int().min(0).max(1000),
  singularityRisk: z.number().int().min(0).max(1000),
}).refine((w) => w.axisTravel + w.limitPenalty + w.singularityRisk > 0, "at least one weight must be positive");
export type SolutionWeights = z.infer<typeof SolutionWeightsSchema>;

export const MachineAxisPositionSchema = z.discriminatedUnion("kind", [
  z.strictObject({ axisId: UuidSchema, kind: z.literal("linear"), positionMm: FiniteNumberSchema }),
  z.strictObject({ axisId: UuidSchema, kind: z.literal("rotary"), positionRad: FiniteNumberSchema }),
]);

/** Captures a validated copy so later caller mutation cannot change validation. */
export function createMachinePluginStateSchema(definition: unknown) {
  const plugin = MachinePluginSchema.parse(definition);
  const axes = new Map(plugin.machine.axes.map((axis) => [axis.id, axis]));
  return z.strictObject({
    contractVersion: z.literal(1),
    pluginId: z.literal(plugin.pluginId),
    pluginVersion: z.literal(plugin.pluginVersion),
    machineId: z.literal(plugin.machine.id),
    mode: FiveAxisModeSchema,
    tcpEnabled: z.boolean(),
    positions: z.array(MachineAxisPositionSchema).length(5),
  }).superRefine((state, context) => {
    const issue = (path: (string | number)[], message: string) =>
      context.addIssue({ code: "custom", path, message });
    if (!plugin.capabilities.modes.includes(state.mode)) {
      issue(["mode"], "mode is not supported by the plugin");
    }
    if (state.tcpEnabled && plugin.capabilities.tcp !== "supported") {
      issue(["tcpEnabled"], "TCP is not supported by the plugin");
    }
    const seen = new Set<string>();
    state.positions.forEach((position, index) => {
      const axis = axes.get(position.axisId);
      if (!axis || seen.has(position.axisId) || axis.kind !== position.kind) {
        issue(["positions", index], "axis must be known, unique and use the correct unit");
      } else {
        const outside = axis.kind === "linear" && position.kind === "linear"
          ? position.positionMm < axis.minMm || position.positionMm > axis.maxMm
          : axis.kind === "rotary" && position.kind === "rotary"
            ? position.positionRad < axis.minRad || position.positionRad > axis.maxRad
            : true;
        if (outside) issue(["positions", index], "axis position exceeds inclusive travel limits");
      }
      seen.add(position.axisId);
    });
  });
}

File: crates/cnc-render-contracts/src/machine_plugin.rs
use crate::domain::{
    self, ContractValidate, DirectionUnit, KinematicAxis, MachineDefinition, MachineType,
    Transform, Vec3Mm,
};
use crate::{ContractError, ContractResult};
use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, BTreeSet};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Architecture {
    TableTable,
    HeadTable,
    HeadHead,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum FiveAxisMode {
    #[serde(rename = "3plus2")]
    Indexed,
    #[serde(rename = "simultaneous-5axis")]
    Simultaneous,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum TcpSupport {
    Supported,
    Unsupported,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Capabilities {
    pub modes: Vec<FiveAxisMode>,
    pub tcp: TcpSupport,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ToolMount {
    pub position_mm: Vec3Mm,
    pub tool_axis_unit: DirectionUnit,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct FiveAxisTarget {
    pub tcp_position_mm: Vec3Mm,
    pub tool_axis_unit: DirectionUnit,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RewindDirection {
    Positive,
    Negative,
}
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct RewindRequest {
    pub axis_id: String,
    pub direction: RewindDirection,
    pub retract_distance_mm: f64,
}
impl ContractValidate for RewindRequest {
    fn validate(&self) -> ContractResult<()> {
        domain::validate_uuid(&self.axis_id, "$.axisId")?;
        require(
            self.retract_distance_mm.is_finite()
                && self.retract_distance_mm > 0.0
                && self.retract_distance_mm <= 500.0,
            "retract distance must be in (0,500] mm",
        )
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SolutionWeights {
    pub axis_travel: u32,
    pub limit_penalty: u32,
    pub singularity_risk: u32,
}
impl Default for SolutionWeights {
    fn default() -> Self {
        Self {
            axis_travel: 10,
            limit_penalty: 1,
            singularity_risk: 2,
        }
    }
}
impl ContractValidate for SolutionWeights {
    fn validate(&self) -> ContractResult<()> {
        require(
            self.axis_travel <= 1000 && self.limit_penalty <= 1000 && self.singularity_risk <= 1000,
            "weights must be integers in 0..=1000",
        )?;
        require(
            self.axis_travel + self.limit_penalty + self.singularity_risk > 0,
            "at least one weight must be positive",
        )
    }
}
impl ContractValidate for FiveAxisTarget {
    fn validate(&self) -> ContractResult<()> {
        domain::validate_vec3(&self.tcp_position_mm, "$.tcpPositionMm")?;
        domain::validate_direction(&self.tool_axis_unit, "$.toolAxisUnit")
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct MachinePlugin {
    pub contract_version: u32,
    pub plugin_id: String,
    pub plugin_version: String,
    pub architecture: Architecture,
    pub machine: MachineDefinition,
    pub capabilities: Capabilities,
    pub tool_chain_axis_ids: Vec<String>,
    pub workpiece_chain_axis_ids: Vec<String>,
    pub tool_mount: ToolMount,
    pub workpiece_mount: Transform,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(
    tag = "kind",
    rename_all = "lowercase",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum AxisPosition {
    Linear { axis_id: String, position_mm: f64 },
    Rotary { axis_id: String, position_rad: f64 },
}
impl AxisPosition {
    pub fn id(&self) -> &str {
        match self {
            Self::Linear { axis_id, .. } | Self::Rotary { axis_id, .. } => axis_id,
        }
    }
    pub fn value(&self) -> f64 {
        match self {
            Self::Linear { position_mm, .. } => *position_mm,
            Self::Rotary { position_rad, .. } => *position_rad,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct MachinePluginState {
    pub contract_version: u32,
    pub plugin_id: String,
    pub plugin_version: String,
    pub machine_id: String,
    pub mode: FiveAxisMode,
    pub tcp_enabled: bool,
    pub positions: Vec<AxisPosition>,
}

fn require(condition: bool, message: &str) -> ContractResult<()> {
    if condition {
        Ok(())
    } else {
        Err(ContractError {
            code: "machine-plugin.invalid",
            path: "$".into(),
            message: message.into(),
        })
    }
}

impl ContractValidate for MachinePlugin {
    fn validate(&self) -> ContractResult<()> {
        require(self.contract_version == 1, "unsupported contract version")?;
        require(
            !self.plugin_id.is_empty()
                && self.plugin_id.len() <= 128
                && self.plugin_id.split(['.', '-']).all(|part| {
                    !part.is_empty()
                        && part
                            .bytes()
                            .all(|b| b.is_ascii_lowercase() || b.is_ascii_digit())
                }),
            "invalid plugin ID",
        )?;
        let version: Vec<_> = self.plugin_version.split('.').collect();
        require(
            self.plugin_version.len() <= 32
                && version.len() == 3
                && version
                    .iter()
                    .all(|part| !part.is_empty() && part.bytes().all(|b| b.is_ascii_digit())),
            "invalid plugin version",
        )?;
        domain::validate_machine(&self.machine, "$.machine")?;
        domain::validate_vec3(&self.tool_mount.position_mm, "$.toolMount.positionMm")?;
        domain::validate_direction(&self.tool_mount.tool_axis_unit, "$.toolMount.toolAxisUnit")?;
        domain::validate_transform(&self.workpiece_mount, "$.workpieceMount")?;
        require(
            matches!(
                self.machine.machine_type,
                MachineType::VerticalMachiningCenter | MachineType::HorizontalMachiningCenter
            ),
            "machining center required",
        )?;
        require(
            self.machine.axes.len() == 5
                && self
                    .machine
                    .axes
                    .iter()
                    .filter(|a| matches!(a, KinematicAxis::Rotary { .. }))
                    .count()
                    == 2,
            "three linear and two rotary axes required",
        )?;
        require(
            !self.capabilities.modes.is_empty()
                && self.capabilities.modes.len() <= 2
                && (self.capabilities.modes.len() == 1
                    || self.capabilities.modes[0] != self.capabilities.modes[1]),
            "invalid modes",
        )?;
        let axes: BTreeMap<_, _> = self.machine.axes.iter().map(|a| (a.id(), a)).collect();
        require(axes.len() == 5, "duplicate axes")?;
        let mut used = BTreeSet::new();
        for chain in [&self.tool_chain_axis_ids, &self.workpiece_chain_axis_ids] {
            require(chain.len() <= 5, "chain too long")?;
            let mut parent = None;
            for id in chain {
                let axis = axes.get(id.as_str()).ok_or_else(|| ContractError {
                    code: "machine-plugin.invalid",
                    path: "$".into(),
                    message: "unknown axis".into(),
                })?;
                require(
                    used.insert(id.as_str()) && axis.parent_id() == parent,
                    "invalid chain",
                )?;
                parent = Some(id.as_str());
            }
        }
        require(used.len() == 5, "missing axis")?;
        let roots: BTreeSet<_> = self
            .machine
            .axes
            .iter()
            .filter(|a| a.parent_id().is_none())
            .map(|a| a.id())
            .collect();
        let declared: BTreeSet<_> = self
            .machine
            .kinematic_root_axis_ids
            .iter()
            .map(String::as_str)
            .collect();
        require(
            roots == declared && declared.len() == self.machine.kinematic_root_axis_ids.len(),
            "invalid roots",
        )?;
        let heads = self
            .tool_chain_axis_ids
            .iter()
            .filter(|id| matches!(axes[id.as_str()], KinematicAxis::Rotary { .. }))
            .count();
        require(
            heads
                == match self.architecture {
                    Architecture::TableTable => 0,
                    Architecture::HeadTable => 1,
                    Architecture::HeadHead => 2,
                },
            "architecture mismatch",
        )?;
        let min = &self.machine.work_envelope.min_mm;
        let max = &self.machine.work_envelope.max_mm;
        require(
            min.x_mm < max.x_mm && min.y_mm < max.y_mm && min.z_mm < max.z_mm,
            "invalid envelope",
        )
    }
}

impl MachinePluginState {
    pub fn validate_for(&self, plugin: &MachinePlugin) -> ContractResult<()> {
        require(
            self.contract_version == 1
                && self.plugin_id == plugin.plugin_id
                && self.plugin_version == plugin.plugin_version
                && self.machine_id == plugin.machine.id,
            "state identity mismatch",
        )?;
        require(
            plugin.capabilities.modes.contains(&self.mode)
                && (!self.tcp_enabled || plugin.capabilities.tcp == TcpSupport::Supported),
            "unsupported mode or TCP",
        )?;
        require(self.positions.len() == 5, "five positions required")?;
        let mut used = BTreeSet::new();
        for position in &self.positions {
            let value = position.value();
            require(
                value.is_finite()
                    && !(value == 0.0 && value.is_sign_negative())
                    && used.insert(position.id()),
                "invalid or duplicate position",
            )?;
            let valid = match (
                plugin.machine.axes.iter().find(|a| a.id() == position.id()),
                position,
            ) {
                (
                    Some(KinematicAxis::Linear { min_mm, max_mm, .. }),
                    AxisPosition::Linear { .. },
                ) => value >= *min_mm && value <= *max_mm,
                (
                    Some(KinematicAxis::Rotary {
                        min_rad, max_rad, ..
                    }),
                    AxisPosition::Rotary { .. },
                ) => value >= *min_rad && value <= *max_rad,
                _ => false,
            };
            require(valid, "unknown axis, wrong unit or travel exceeded")?;
        }
        Ok(())
    }
}
